metrics/firebase/functions/test/main_test.dart

```diff
@@ -450,6 +454,441 @@ void main() {
       },
     );
   });
+
+  group("onBuildUpdatedHandler", () {
+    final newDocumentSnapshot = DocumentSnapshotMock();
+    final change = ChangeMock();
+    final documentQuery = DocumentQueryMock();
+    final querySnapshot = QuerySnapshotMock();
+    final oldBuildJson = testDataGenerator.generateBuildJson(
+      buildStatus: BuildStatus.inProgress,
+      duration: null,
+    );
+    final oldBuildDocumentData = DocumentData.fromMap(oldBuildJson);
+
+    PostExpectation<Future<QuerySnapshot>> whenGetTaskDocument() {
+      whenFirestore().thenReturn(firestore);
+      when(firestore.collection(tasksCollectionName))
+          .thenReturn(collectionReference);
+      when(collectionReference.where(
+        any,
+        isEqualTo: anyNamed('isEqualTo'),
+      )).thenReturn(documentQuery);
+
+      return when(documentQuery.get());
+    }
+
+    PostExpectation<List<DocumentSnapshot>> whenTaskDocuments() {
+      whenGetTaskDocument().thenAnswer((_) => Future.value(querySnapshot));
+
+      return when(querySnapshot.documents);
+    }
+
+    PostExpectation<DocumentData> whenAccessChangeData({
+      DocumentSnapshot documentSnapshot,
+      bool isAfter = true,
+    }) {
+      final PostExpectation whenChange =
+          isAfter ? when(change.after) : when(change.before);
+
+      whenChange.thenReturn(documentSnapshot);
+
+      return when(documentSnapshot.data);
+    }
+
+    setUp(() {
+      whenAccessChangeData(
+        documentSnapshot: documentSnapshot,
+        isAfter: false,
+      ).thenReturn(oldBuildDocumentData);
+      whenDocument().thenReturn(documentReference);
+      whenTaskDocuments().thenReturn([]);
```

Comment:
I believe it would be better to avoid mocking any data in `setUp` methods since it may make tests not clear and too complicated.